### ISSUE TITLE
Secure Credentials Manager bug decrypting Credentials within the SDK

### ISSUE THREAD

**Issue by davgdafa** [auth0/Auth0.Android#330]
### Description

Even though the newest version of the Credentials Manager, Secure Credentials Manager, is currently refreshing the token in our Android app, it is presenting problems that it invalidates the credentials because of an error decrypting the credentials **when app is closed and reopened**.

This happens in the Auth0 SDK latest version 1.24.1 and 1.20.0 (com.auth0.android:auth0) at SecureCredentialsManager.java (Line 250: “A change on the Lock Screen security settings have deemed the encryption keys invalid and have been recreated. "; "Any previously stored content is now lost. Please, try saving the credentials again.”).
 
I tracked this down to the next class called CryptoUtil.java, Line 436: “The AES encrypted input is corrupted and cannot be recovered. Please discard it.” -> BadPaddingException: “error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT”.
 
This seems to be a problem with the padding when writing this to local memory within the SDK and this seems to be happening on Android OS 6 and newer according to some posts in the internet doing a search over “error:1e000065:Cipher functions:OPENSSL_internal:BAD_DECRYPT”.

This below is our implementation of the Universal Login:
```
WebAuthProvider.login(OUR_AUTH0_ACCOUNT)
                .withScheme(OUR_SCHEME)
                .withScope("openid offline_access")
                .withAudience(OUR_AUDIENCE)
                .start(activity, object : AuthCallback {
                    override fun onFailure(dialog: Dialog) {
                        emitter.onError(Throwable("signInWithAuth0 login onFailure and dialog:"))
                    }

                    override fun onFailure(exception: AuthenticationException?) {
                        emitter.onError(Throwable("signInWithAuth0 login onFailure and dialog:"))
                    }

                    override fun onSuccess(credentials: Credentials) {
                        emitter.onSuccess(credentials)
                    }
                })
```

### Reproduction

1) Login using the Universal Login web form (launched using code above)
2) **Go to another Activity (screen), kill the app. Either by pressing menu and closing it or letting the OS kill it after a while.**
3) Check the SecureCredentialsManager(...).hasValidCredentials() to make an API call to get some data using the Auth0 token.
**Expected**: User has valid credentials and we make the api call.
**Actual**: Secure Credentials Managers throws an error saying: 
SecureCredentialsManager.java Line 250: “A change on the Lock Screen security settings have deemed the encryption keys invalid and have been recreated. "; "Any previously stored content is now lost. Please, try saving the credentials again."

Test was made in an Android emulator from Android Studio with these settings:
![image](https://user-images.githubusercontent.com/24528079/90916161-99100500-e3ae-11ea-86e2-33ba11c7dfcd.png)

This is working using the CredentialsManager, but we want to use the latest version of it since it is more secure.

### Environment

- **Version of this library used: tested with 1.24.1 and 1.20.0**
- **Version of the platform or framework used, if applicable: Android OS 10.0**
- **Other relevant versions (language, server software, OS, browser): Kotlin**
- **Other modules/plugins/libraries that might be involved: Koin DI**

**Comment by davgdafa** [comment]
@lbalmaceda Hi Luciano, I wanted to make sure my steps to reproduce were actually the ones.

I replicated the bug every time I restarted the app (after being previously authenticated):
1) Login using the Universal Login web form (launched using code above)
2) **Go to another Activity (screen), kill the app. Either by pressing menu and closing it or letting the OS kill it after a while.**
3) Check the SecureCredentialsManager(...).hasValidCredentials() to make an API call to get some data using the Auth0 token.
**Expected**: User has valid credentials and we make the api call.
**Actual**: Secure Credentials Managers throws an error saying:
SecureCredentialsManager.java Line 250: “A change on the Lock Screen security settings have deemed the encryption keys invalid and have been recreated. "; "Any previously stored content is now lost. Please, try saving the credentials again."

Therefore, I am actually able to navigate through our app, making API calls perfectly. When I close the app, or reopen it, the SecureCredentialsManager gets that error I have in my original comment.

Hopefully, that helps.

**Comment by lbalmaceda** [comment]
Thanks again for the steps. So far, I was not able to reproduce this without DI. 

I've created a [sample app in Kotlin](https://github.com/lbalmaceda/ReproduceableSamplesAuth0/tree/master/SecureCredentialsManager) that does a basic web auth flow and allows the user to get to a second activity, provided the `hasValidCredentials` returns true. Feel free to clone the sample and try to run that locally.

I'm loading the credentials on that second activity for displaying/testing purposes. Then I go to a different app, close the sample app activity, and try to reopen. The credentials are there. The `hasValidCredentials` still returns true, and I can access them without a problem. Please, if I'm not following the exact repro steps you mentioned, let me know. 

Here is a GIF that lasts approx 1 min. 
![kapture](https://user-images.githubusercontent.com/3900123/91331786-9f80f100-e7a1-11ea-846e-fb0b838e19b9.gif)

And the AVD configuration
![image](https://user-images.githubusercontent.com/3900123/91332202-351c8080-e7a2-11ea-9d1b-9ba860ef9831.png)

**Comment by davgdafa** [comment]
@lbalmaceda Awesome, thanks for the quick response. I'll test and let you know 👍

**Comment by davgdafa** [comment]
@lbalmaceda:

I tried your sample and it worked great.

After a good debugging session of the `SecureCredentialsManager` and the `CryptoUtils` in our project, I found what's below:

The problem in our project _was_ when we were calling the `SecureCredentialsManager.saveCredentials()`; we were saving credentials several times when we finished logging in, and every time we got the Credentials from the SecureCredentialsManager; this was invalidating the credentials because the IV code changes to encrypt/decrypt and we are doing simultaneous API calls. Therefore, at some point the storage had a KEY_IV_ALIAS value different, and finally, AES would not decrypt the values and invalidate the credentials. That's why they were invalid. I also found another bug in our app that when credentials were invalidated, we weren't signing out the user in our app, and the user seemed as "Signed In". So, no need to even close and open the app; it would just in the next iteration of retrieving the Credentials.

I solved it by only saving the new credentials when we login. I did notice this is done internally (in the SecureCredentialsManager), when the pair of tokens are renewed.

I'm finally testing that the renew() is working correctly when the access token expires, so in about two hours, when the access token expires, I will debug to see if everything looks correct.

Thank you so much.

P.S. It would be nice to have that in the documentation. I really haven't found the specific documentation to this Auth0.Android SDK, but if you have one, please point me to the right one for best practices 👍 🥇

**Comment by davgdafa** [comment]
So, I tested that the SecureCredentialsManager is in fact refreshing the token without invalidating the credentials. 

My findings are divided in two scenarios:

**Scenario "Token expires while App is not running" - Should renew the tokens: PASSED**

A user authenticates using Auth0 and we save the Credentials using the `SecureCredentialsManager`, and we get a refresh token as well. We close the app. We, then wait for the tokens to expire.

When the app starts, we go through an async process, but we wait for the authentication first, if authentication succeeds, then we go ahead and we make API calls; if not, the user is a guest and go to a different route. In this case, we check only if the credentials are valid `hasValidCredentials()` and returns true because the refresh token is stored.

When we "getCredentials()" an API call is triggered to renew the pair of tokens (all remotely through the SDK). 

In this case, we are good 👍 Credentials are retrieved.

**Scenario "Token expired while User has App open" - Should renew the tokens: FAILS**

A user authenticates using Auth0 and we save the Credentials using the `SecureCredentialsManager`, and we get a refresh token as well. We keep app open and phone awake (emulator).

We attempt to refresh the homepage which triggers parallel API calls. The pair of token is refreshed three times remotely, ultimately obtaining the same pair of tokens; however, the fourth time, the Credentials are invalidated because of the same issue, the IV has changed because the `saveCredentials` operation has been done simultaneously.

In this last case, the test fails. User is not authenticated anymore and we go through the guest user route.

What do you recommend to avoid this from happening?

**Comment by lbalmaceda** [comment]
@davgdafa thanks for sharing that! I'm happy that you were able to fix the original issue.

> P.S. It would be nice to have that in the documentation. I really haven't found the specific documentation to this Auth0.Android SDK, but if you have one, please point me to the right one for best practices 👍 🥇

All the documentation we have for the Credential Managers (both) is in the [readme](https://github.com/auth0/Auth0.Android#credentials-manager) and on the [site](https://auth0.com/docs/libraries/auth0-android/auth0-android-save-and-renew-tokens). Was there anything in particular that you couldn't find and we could add? or maybe something that is already documented but poorly explained?

> I solved it by only saving the new credentials when we login.

Typically, the user would be asked to log in once at the app first launch. And then, if you have a refresh token, you could keep that session valid until the user decides to log out. Yes, getting the saved credentials would be an async process because if they have expired, the manager would need to perform an HTTP request to refresh them. You should wait until this completes before moving to a section of your app that will use the tokens. 

You did mention saving credentials multiple times. Out of curiosity, can you share more details on that implementation? 



> Scenario "Token expired while User has App open" - Should renew the tokens: FAILS

Always get the tokens from the `getCredentials` method. If you do that once, say in "onCreate", and store that in a class variable / property, you risk of falling in this scenario where the user might hide the app and come back later when that token is no longer valid. If you always access the tokens from a fresh call to `getCredentials`, you always get a valid token, because the manager will refresh it if it has expired (provided a refresh token was saved originally, it will throw otherwise).

> The pair of token is refreshed three times remotely

This doesn't look right. Refreshing it once should be enough, since the new received access token will probably last at least an entire day by default (depends on your tenant configuration, anyway). Are you using multiple instances of the manager or are you calling `getCredentials` multiple times at approximately the same moment? Notice in my sample I'm not even using a singleton nor DI, but that's just because it was a sample app. I'd try to use a singleton.

**Comment by davgdafa** [comment]
@lbalmaceda thank you for your quick response

> All the documentation we have for the Credential Managers (both) is in the readme and on the site. Was there anything in particular that you couldn't find and we could add? or maybe something that is already documented but poorly explained?

Yes. It would be good to state that if the `SecureCredentialsManager` is used, the method `savedCredentials()` can only be used once after successful login. In the other hand, when `getCredentials()` is used, like you said, if needed, it will automatically renew the pair of tokens synchronously (it will wait until it retrieves them even if it's done with a HTTP request), and therefore, can't be called simultaneously and asynchronously because if the pair of tokens expires, when renewing, it ultimately can cause the SDK to invalidate the credentials because of the AES encryption.

 > You should wait until this completes before moving to a section of your app that will use the tokens.

Ultimately, this would solve the problem. In our app, we only wait for the credentials in the App start; we currently have screens **_that require simultaneous API calls_** to load the data. Every single time we make an API call, we first get the token using the `getCredentials` method, so in an example where we do 2 API calls simultaneously, the process is going to be happening twice (renew pair of tokens), and that would break that concept of waiting for it. We would have to change the implementation to sequential API calls instead and this would affect our Production code.

> are you calling getCredentials multiple times at approximately the same moment?

Yep. I think that's the problem. Originally, our implementation of the SecureCredentialsManager was a Singleton by definition (single) but I changed it to factory (which would create a new one every time), though in my last testing, and for the sake of testing if DI was the problem I switched to `factory` and also tested creating them every time called.

I'll change back to singleton to see if the behavior improves, but I do not expect it to improve since we are kind of breaking the concept of waiting for it to make the next API call, and instead we are making several API calls at the same time.

**Comment by lbalmaceda** [comment]
> The method savedCredentials() can only be used once after successful login

No, that's not accurate. You could save credentials multiple times, but the user in a **typical** authentication flow is asked to log in once at launch time or in some other part of the app, and only if they were not already logged in. If you are saving the obtained credentials there, where else in your app do you need to save other credentials again? That's where I no longer follow you.

> Every single time we make an API call, we first get the token using the getCredentials method

I think in this case what would have helped to diagnose this earlier is mentioning that the classes are **not thread-safe**. Because of your multiple requests scenario, you first need to ensure you have a valid token, and only when that value is there, you can start the train of API calls, which **can** be parallel.

The problem, as you mentioned in a previous comment, seems to be around the async nature of the class. When you retrieve the credentials for one API call and they have expired, it will kick a refresh call which will end up writing the new values to the storage. At the same time, you start a second API call that will do the same. Eventually, at some point, the `CryptoUtil` can't handle the asynchronism and breaks its state. Sort of like the "readers-writers problem" or a race condition issue. This is because the decryption process is a two-step process, and two related values need to be fetched from the store before the value can be decrypted. I suspect one of them changes, making the other no longer be valid and failing with the ultimate removal of the well-stored credentials from the storage.


In terms of documentation, we are [already covering](https://github.com/auth0/Auth0.Android#credentials-manager) the class behavior when a refresh token is present. Because of the interaction with the authentication API, in both classes the `getCredentials` methods are async. What we can clarify is that race condition behavior when the token expires and need to be refreshed. What do you think?

**Comment by davgdafa** [comment]
We were having issues at the beginning trying to renew the credentials because we weren't passing "offline_access" in the scope and that was in the documentation; that's why I started saving credentials more than once, to catch a possible updated credential. Please ignore why I did it, I just tried a bunch of stuff originally to make it work using the documentation and didn't catch that.

I was able to test as Singleton classes (SecureCredentialsManager) and when returned to the app after expired, it did do 4 API calls which is totally expected, and it did not invalidate the credentials, which will have to do this test several times to say that this is not a problem in the SDK for us.

Overall, we will be switching to `SecureCredentialsManager` because our main concern is the security part. 

I thank you for all your support on this issue; I believe at this point the original issue can be closed.

**Comment by lbalmaceda** [comment]
Still, those repeated refresh token requests should be optimized. For that to happen internally, the SDK would need to handle that thread concurrency, which I'm not 100% sure how it will play on an Android app, or have some sort of token cache. I'll add a card to discover this in the coming weeks.

Let me try to rephrase the whole credentials manager section of the readme and at least explain better the refresh token + offline access part. Stay tuned for a PR, I'll keep this issue open until that's posted 👍

**Comment by poovamraj** [comment]
Hi everyone, if you are still looking for a thread-safe version of this method, we have released it in [2.7.0](https://github.com/auth0/Auth0.Android/releases/tag/2.7.0). Do take a look.
